clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -7540,6 +7540,17 @@ static bool evaluateAddIRAttributesArgs(Expr **ArgsBegin, size_t ArgsSize,
   return false;
 }
 
+SYCLAddIRAttributesFunctionAttr *Sema::MergeSYCLAddIRAttributesFunctionAttr(
+    Decl *D, const SYCLAddIRAttributesFunctionAttr &A) {
+  if (const auto *ExistingAttr =
+          D->getAttr<SYCLAddIRAttributesFunctionAttr>()) {
+    Diag(ExistingAttr->getLoc(), diag::warn_duplicate_attribute_exact) << &A;
+    Diag(A.getLoc(), diag::note_previous_attribute);
+    return nullptr;
+  }
+  return A.clone(Context);
```

Comment:
Are we ignoring duplicate attribute values here? could you please add tests for it?

Reply:
I have added the test cases. It does generate the same diagnostics as when the attributes are not the exact same.

Reply:
It is expected to inherit the attribute from the declaration, otherwise I don't think the warning we produce for duplicate attributes in redeclarations make much sense either. I have added a test case for both `add_ir_attributes_function` and `add_ir_attributes_global_variable` in clang/test/SemaSYCL/attr-add-ir-attributes-merge.cpp.

Reply:
This has been discussed offline and is covered by testing. Redeclarations will inherit attributes and redeclarations with the same attribute as a previous declaration in the chain need to have the same arguments (potentially in different orders).